app/models/foreman_tasks/task/dynflow_task.rb

```diff
@@ -19,6 +19,7 @@ def update_from_dynflow(data)
                                 end
                               end
       self.label          ||= main_action && main_action.class.name
+      self.action = format_input(true)
```

Comment:
Wouldn't it be better to put action calculation to `:after_validate` callback? This way we will be sure it's updated each time we save the model.

Reply:
Would work for me. Also: would it work if we did `self.action ||= `, to save some calcutlation time on each save?

Reply:
> would it work if we did self.action ||=, to save some calcutlation time on each save?

It depends if `:name` and `:input` could be changed. For clarity I would suggest using condition on the callback:
``` ruby
after_validation :set_action_field :if => Proc.new { |task| task.name_changed? || task.input_changed? }
```

Reply:
The thing here is that `name` and `input` are not attributes of the model so we're not getting all the `*_changed?` goodies from rails. Not sure about `name`, but `input` can definitely change during the run of the task so `||=` is probably not a good idea. Any other idea than implementing `*_changed?` ourselves?

Reply:
If we can't track changes to `name` and `input`, we will not be able to avoid recalculation. We don't have to write field tracking code from scratch, since we already have it in our models from [ActiveMode::Dirty](http://api.rubyonrails.org/classes/ActiveModel/Dirty.html). Anyway, the field will be omitted from SQL UPDATE statement, if the value remains the same (tested it on another ActiveRecord). I don't think the calculation is that expensive that we should try and optimize it. IMHO, we can set the field each save without any worries.